pkg/kube/events.go

```diff
@@ -0,0 +1,115 @@
+package kube
+
+import (
+	"github.com/golang/glog"
+	"github.com/skupperproject/skupper/api/types"
+	"github.com/skupperproject/skupper/pkg/event"
+	v1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	typedcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/record"
+)
+
+const (
+	EventRecorderRoleName        string = "event-recorder"
+	EventRecorderRoleBindingName string = "event-recorder"
+)
+
+var EventRecorderPolicyRule = []rbacv1.PolicyRule{
+	{
+		Verbs:     []string{"watch", "create", "patch"},
+		APIGroups: []string{""},
+		Resources: []string{"events"},
+	},
+}
+
+type SkupperEventRecorder struct {
+	EventRecorder record.EventRecorder
+	Source        *v1.Service
+	Disabled      bool
+}
+
+func NewEventRecorder(namespace string, cli kubernetes.Interface, disabled bool) SkupperEventRecorder {
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.Infof)
+	eventBroadcaster.StartRecordingToSink(
+		&typedcorev1.EventSinkImpl{
+			Interface: cli.CoreV1().Events(namespace)})
+	kubeEventRecorder := eventBroadcaster.NewRecorder(
+		scheme.Scheme,
+		v1.EventSource{
+			Component: types.ControllerServiceName})
+	service, _ := cli.CoreV1().Services(namespace).Get(types.ControllerServiceName, metav1.GetOptions{})
+
+	eventRecorder := SkupperEventRecorder{
+		EventRecorder: kubeEventRecorder,
+		Source:        service,
+		Disabled:      disabled,
+	}
+	return eventRecorder
+}
+
+func RecordWarningEvent(reason string, message string, recorder SkupperEventRecorder) {
```

Comment:
What if we defined an interface under pkg/event with RecordWarning/RecordNormal functions, and then could create two implementations of that, one that just does what we do now, and one that records the event through kube as well, each with a distinct factory function.

Reply:
Another kinda related thing. Why not making `RecordWarningEvent` and `RecordNormalEvent` methods of `SkupperEventRecorder` ? That would eliminate the need of passing the SkupperEventRecorder instance
every time.

Reply:
Agreed, that was implied by having an interface.